libdevice/fallback-cstring.cpp

```diff
@@ -7,9 +7,89 @@
 //===----------------------------------------------------------------------===//
 
 #include "wrapper.h"
+#include <cstdint>
 
 #ifdef __SPIR__
+DEVICE_EXTERN_C
 void *__devicelib_memcpy(void *dest, const void *src, size_t n) {
   return __builtin_memcpy(dest, src, n);
 }
+
+DEVICE_EXTERN_C
+void *__devicelib_memset(void *dest, int c, size_t n) {
+  return __builtin_memset(dest, c, n);
+}
+
+static int __devicelib_memcmp_1byte(const void *s1, const void *s2, size_t n) {
+  const uint8_t *s1_uint8_ptr = reinterpret_cast<const uint8_t *>(s1);
+  const uint8_t *s2_uint8_ptr = reinterpret_cast<const uint8_t *>(s2);
+  while (n > 0) {
+    if (*s1_uint8_ptr == *s2_uint8_ptr) {
+      s1_uint8_ptr++;
+      s2_uint8_ptr++;
+      n--;
+    } else {
+      return *s1_uint8_ptr - *s2_uint8_ptr;
+    }
+  }
+
+  return 0;
+}
+
+static int __devicelib_memcmp_4byte(const void *s1, const void *s2, size_t n) {
+  const uint32_t *s1_uint32_ptr = reinterpret_cast<const uint32_t *>(s1);
+  const uint32_t *s2_uint32_ptr = reinterpret_cast<const uint32_t *>(s2);
+  while (n >= 4) {
+    if (*s1_uint32_ptr == *s2_uint32_ptr) {
+      s1_uint32_ptr++;
+      s2_uint32_ptr++;
+      n -= 4;
+    } else {
+      n = 4;
+      break;
+    }
+  }
+
+  const uint8_t *s1_uint8_ptr =
+      reinterpret_cast<const uint8_t *>(s1_uint32_ptr);
+  const uint8_t *s2_uint8_ptr =
+      reinterpret_cast<const uint8_t *>(s2_uint32_ptr);
+
+  for (size_t idx = 0; idx < n; ++idx) {
+    if (s1_uint8_ptr[idx] == s2_uint8_ptr[idx])
+      continue;
+    else
+      return s1_uint8_ptr[idx] - s2_uint8_ptr[idx];
+  }
+
+  return 0;
+}
+
+DEVICE_EXTERN_C
+int __devicelib_memcmp(const void *s1, const void *s2, size_t n) {
+  if (s1 == s2 || n == 0)
+    return 0;
+
+  size_t s1_4mod = reinterpret_cast<unsigned long>(s1) % 4;
```

Comment:
The most portable way is to use `alignof(uint32_t)` instead of `4`.  I would also recommend renaming `__devicelib_memcmp_4byte` to something like `__devicelib_memcmp_uint32_aligned`.

Reply:
Hi, @vzakhari 
Already replaced all hardcode with alignof and sizeof and corresponding function name has been updated.
Thanks very much.